### DISCUSSION TITLE
Global Lifetime Does Not Execute Task Lifetime

### DISCUSSION THREAD

**Discussion by BlythMeister** [cake-build/cake#3150] [Q&A]
### What You Are Seeing?

Global Lifetime invocation does not trigger Task Lifetime

Net result, when setting progress blocks for TeamCity in Task Lifetimes (setup and teardown) these do not appear in TeamCity log for global setup/teardown

I have attempted to put the TeamCity block logging into my setup/teardown but the header does not get put into that block.

As an asside (pure observation) the logging of the titles for global setup/teardown are a different format to tasks.

### What is Expected?

Execution order

Task Lifetime Setup
Global Lifetime Setup
Task Lifetime Teardown

.... OTHER TARGETS ....

Task Lifetime Setup
Global Lifetime Teardown
Task Lifetime Teardown

### What version of Cake are you using?

V1 (With Fosting)

### Are you running on a 32 or 64 bit system?

64 bit

### What environment are you running on?  Windows? Linux? Mac?

Windows

### Are you running on a CI Server?  If so, which one?

TeamCity

### How Did You Get This To Happen? (Steps to Reproduce)

Simple setup with no targets, but a task and global lifetime.

Task Lifetime Setup has:

```
            if (context.TeamCity().IsRunningOnTeamCity)
            {
                context.TeamCity().WriteStartProgress(info.Task.Name);
            }
```

Task Lifetime Teardown has:

```
            if (context.TeamCity().IsRunningOnTeamCity)
            {
                context.TeamCity().WriteEndProgress(info.Task.Name);
            }
```

Setup/Teardown just logs

### Output Log

~~~sh
Registering task: Default
Registering setup: BuildSetup
Registering teardown: BuildTeardown
Registering task setup: TaskSetup
Registering task setup: TaskTeardown

----------------------------------------
Setup
----------------------------------------
Executing custom setup action...
MY LOG SETUP
Executing custom task setup action (Default)...
##teamcity[progressStart 'Default']

========================================
Default
========================================
Executing task: Default
Finished executing task: Default
Executing custom task teardown action (Default)...
##teamcity[progressFinish 'Default']

----------------------------------------
Teardown
----------------------------------------
Executing custom teardown action...
MY LOG TEARDOWN

Task                          Duration
--------------------------------------------------
Setup                         00:00:00.0145173
Default                       00:00:00.0110553
Teardown                      00:00:00.0045078
--------------------------------------------------
Total:                        00:00:00.0300804
~~~

Because the setup/teardown do not run the lifetime, the logging for setup/teardown is not collapsible in the TeamCity Log.

**Comment by augustoproiete** [comment]
@BlythMeister I didn't try this in Frosting, but I use Global lifetime & Task lifetime with Cake scripting all the time, and I would expect the behavior in Frosting to be the same as Cake scripting.

What makes you expect that the sequence be the one that you mentioned?

> 
> Task Lifetime Setup
> Global Lifetime Setup
> Task Lifetime Teardown
> 
> .... OTHER TARGETS ....
> 
> Task Lifetime Setup
> Global Lifetime Teardown
> Task Lifetime Teardown

This is not the sequence that I would expect (also not the sequence that is [documented](https://cakebuild.net/docs/writing-builds/setup-and-teardown)).

The Global lifetime are _outer_ tasks, so they wrap everything else.
- Setup runs once, BEFORE the *FIRST* task
- Teardown runs once, AFTER the *LAST* task
- TaskSetup runs once per task, BEFORE *EACH* task
- TaskTeardown runs once per task, AFTER *EACH* task

Thus the sequence you should get (unless you found a bug in Frosting) should be:

```
(1) Setup
    (2) TaskSetup
        (3) TaskA
    (4) TaskTeardown
    (5) TaskSetup
        (6) TaskB
    (7) TaskTeardown
    // ... (repeat for other tasks)
(8) Teardown
```

**Comment by BlythMeister** [comment]
The reason I expected/wanted it is so I can include the logic to write TeamCity service message before and after the logic 

Whilst I could put that inside the global lifetime, this meant the default logging at the start is outside the teamcity block.

I have actually now worked around this by not using global lifetime, and just had a task that runs at the start.
I also realised I never needed the global Teardown.

So, in short.  If this is how it's expected to work...then that's cool.

I wonder if a "pre-run" and "post-run" on global lifetime setup and teardown would enable the custom logging for teamcity?

**Comment by gep13** [comment]
@BlythMeister just to check... have you seen the Cake.BuildSystems.Module?

**Comment by BlythMeister** [comment]
@gep13 OH WOW!!!
No i hadn't see it!
Going to close this, as the only reason i'm doing what i am doing is because of TeamCity :p
Just need to work out how to add modules!

**Comment by gep13** [comment]
@BlythMeister I have been reliably informed that you should be able to reference the Cake.BuildSystems.Module package as a normal package reference, and from there, you should be able to do:

```
services.UseModule<TeamCityModule>()
```

when setting up the services for your Frosting project.

**Comment by BlythMeister** [comment]
That's what I gathered.
Many thanks!
